src/main/perl/Element.pm

```diff
@@ -559,33 +585,62 @@ sub getTree
 
 SWITCH:
     {
+        # LIST to array ref
         $self->isType(LIST) && do {
             $ret = [];
             while ($self->hasNextElement) {
                 $el = $self->getNextElement();
-                push(@$ret, $el->getTree($nextdepth));
+                push(@$ret, $el->getTree($nextdepth, %opts));
             }
             last SWITCH;
         };
+
+        # NLIST to hashref
         $self->isType(NLIST) && do {
             $ret = {};
             while ($self->hasNextElement) {
                 $el = $self->getNextElement();
-                $$ret{$el->getName()} = $el->getTree($nextdepth);
+                $$ret{$el->getName()} = $el->getTree($nextdepth, %opts);
             }
             last SWITCH;
         };
+
+        # BOOLEAN via convert_bool method
         $self->isType(BOOLEAN) && do {
             $ret = $self->getValue eq 'true' ? 1 : 0;
+            $convm = $opts{convert_boolean};
+            last SWITCH;
+        };
+
+        # STRING to string
+        $self->isType(STRING) && do {
+            $ret = $self->getValue;
+            $ret = "$ret";
+            $convm = $opts{convert_string};
+            last SWITCH;
+        };
+
+        # LONG to integer/long
+        $self->isType(LONG) && do {
+            $ret = 0 + $self->getValue;
+            $convm = $opts{convert_long};
+            last SWITCH;
+        };
+
+        # DOUBLE to float/double
+        $self->isType(DOUBLE) && do {
+            $ret = 0.0 + $self->getValue;
+            $convm = $opts{convert_double};
             last SWITCH;
         };
 
-        # Default clause
+        # Default clause, should never be reached
```

Comment:
Throwing an error is generally good but it means that if PAN ever introduces a new data type then data type cannot be used until the node's CCM library is updated to know about it. That's not too unreasonable except that it looks like the current code would handle new types that mapped to Perl scalars just fine anyway, so is the error actually adding any value?

I would be inclined to update the comment instead of throwing an error, e.g. "Default clause: should only be reached if PAN adds a new type which is not explicitly checked above."